src/Elastic.Apm.SmartSql/SmartSqlDiagnosticListener.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Data;
+using Elastic.Apm.Api;
+using Elastic.Apm.DiagnosticSource;
+using Elastic.Apm.Logging;
+using Elastic.Apm.Model;
+using SmartSql.Diagnostics;
+
+namespace Elastic.Apm.SmartSql
+{
+	internal class SmartSqlDiagnosticListener : IDiagnosticListener
+	{
+		private readonly ConcurrentDictionary<Guid, Span> _spans = new ConcurrentDictionary<Guid, Span>();
+
+		public SmartSqlDiagnosticListener(IApmAgent agent) => Logger = agent.Logger?.Scoped(nameof(SmartSqlDiagnosticListener));
+
+		private ScopedLogger Logger { get; }
+
+		public string Name => "SmartSqlDiagnosticListener";
+
+		public void OnCompleted() { }
+
+		public void OnError(Exception error) { }
+
+		public void OnNext(KeyValuePair<string, object> kv)
+		{
+			switch (kv.Key)
+			{
+				case string k when k == SmartSqlDiagnosticListenerExtensions.SMART_SQL_BEFORE_COMMAND_EXECUTER_EXECUTE:
+					if (kv.Value is CommandExecuterExecuteBeforeEventData executerExecuteBeforeEventData)
+					{
+						var newsSpan = Agent.TransactionContainer.Transactions.Value.StartSpanInternal(executerExecuteBeforeEventData.Operation,
+							ApiConstants.TypeDb);
+						newsSpan.Tags.Add("statement",executerExecuteBeforeEventData.ExecutionContext.Request.RealSql);
```

Comment:
`Tags` is preserved for users of the public API - so when users manually trace things. We don't use it for auto instrumentation. I see below you already fill the db related fields - as we do EF Core instrumentation here: https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm.EntityFrameworkCore/EfCoreDiagnosticListener.cs#L45.

What's the intention here with adding the same to `Tags`?

Btw. I think the confusion on the meaning of `Tags` is our fault - already saw multiple people using it for these kinds of thing - I'll make sure we document this better!

Reply:
Thanks @gregkalapos !
Thank you for your comments on the submission code.
I will fix a series of questions raised in the comments in the near future.
1.add the minimal test module 
2.remove the custom tag content.
3.add null check for transaction and make end() called.
This project is great. I plan to put it into production.